### DISCUSSION TITLE
Many-to-many relationship question

### DISCUSSION THREAD

**Discussion by BorisBertelsen** [prisma/prisma#22170] [Q&A] [answered]
### Question

Hello everyone,

I recently became interested in Prisma and is currently working on a little project to learn Prisma. I have a question about many-to-many relationship that I hope someone can help me with.

I would like to add the following to a User (skill, skill rating, team rating)
- Assign multiple Skills on a user, eg.: carpenter, painter, plumming
- Each skill has a rating of 1-5. So he's a 5 star carpenter, 2 star painter and 1 star plumming
- I would also like to rate him as a teamplayer based on the 1-5 rating

I'm not sure if this is the correct way of doing it.
My thought is to create RATING which can go from X - Y and reference to it everywhere I need it:
- As Teamplayer rating
- Rate each skill a person has

```prisma
model User {
  id            Int                @id @default(autoincrement())
  name          String
  skills     SkillsToUser[]
  teamPlayer Ratings[]
  ratingsToUser RatingsToUser[]
}

model Skills {
  id     Int              @id @default(autoincrement())
  name   String
  users  SkillsToUser[]
  rating Ratings          @relation(fields: [id], references: [id])
}

model SkillsToUser {
  user       User      @relation(fields: [userId], references: [id])
  userId     Int
  skills   Skills @relation(fields: [skillsId], references: [id])
  skillsId Int       

  @@id([userId, SkillsToUser])
}


model Ratings {
  id            Int             @id @default(autoincrement())
  rating        Int
  crewType      Skills[]
  user          User?           @relation(fields: [userId], references: [id])
  userId        Int?
  RatingsToUser RatingsToUser[]
}

model RatingsToUser {
  user   User @relation(fields: [userId], references: [id])
  userId Int

  rating   Ratings @relation(fields: [ratingId], references: [id])
  ratingId Int

  @@id([userId, ratingId])
}
```

Furthermore, whenever this is setup correctly I'll try and create a seed file.
I hope my question makes sense.

Bests,
Mike

- OS: Ubuntu
- Database: Postgres
- Node.js version: 18.17.1
- Prisma: 5.6.0

**Comment by ludralph** [comment, marked as answer]
Hi @BorisBertelsen 👋 

It's great to hear that you're exploring Prisma for your project. Many-to-many relationships can indeed be a bit tricky, but Prisma provides a way to handle them effectively.

Based on your schema, it seems like you're on the right track with creating a `Ratings` model that can be referenced for different purposes. However, there are a few things to consider:

1. For the `Skills` model, you have a direct relation to `Ratings` using the `id` field. This might not be the correct approach if you want to have a separate rating for each skill. Instead, you might want to have a `SkillsToUser` join table that includes a rating field to store the skill rating for each user.

2. The `teamPlayer` field in the `User` model is an array of `Ratings`, but it's not clear how this relates to the `Ratings` model. If you want to rate the user as a team player, you might need a separate field that stores this rating, or you could include it in the `RatingsToUser` join table.

3. The `RatingsToUser` model seems to be a join table between `User` and `Ratings`, but it's not clear what the rating field in the `Ratings` model represents. If it's the actual numeric rating, then you might not need a separate `RatingsToUser` model; instead, you could include the rating directly in the `SkillsToUser` join table. 

Based on this observations, you can modify the schema to look like below:
```prisma
model User {
  id         Int            @id @default(autoincrement())
  name       String
  skills     SkillsToUser[]
  teamRating Int // Assuming team rating is a simple integer value
}

model Skill {
  id    Int            @id @default(autoincrement())
  name  String
  users SkillsToUser[]
}

model SkillsToUser {
  user    User  @relation(fields: [userId], references: [id])
  userId  Int
  skill   Skill @relation(fields: [skillId], references: [id])
  skillId Int
  rating  Int // Rating for a specific skill

  @@id([userId, skillId])
}
```

In the schema above, 

- Each `User` has multiple `SkillsToUser`, which represents the many-to-many relationship between User and Skill.
- The `SkillsToUser` join table includes a `rating` field to store the skill rating for each user.
- The `teamRating` field is added directly to the User model to store the team player rating.

As for seeding the database, once you have your schema set up, you can use Prisma's seeding functionality to populate your database with initial data. You can find more information on how to do this in the [Prisma documentation](https://www.prisma.io/docs/guides/database/seed-database).

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by BorisBertelsen** [comment]
Hi @ludralph,

Thank you very much for coming by and providing help. I appreciate it.
I appreciate your explanation and I'd love to test it out by updating my seeds file.

Thank you very much for the help! Fingers crossed on my next adventure ;)
